2024/BIN-2024-0003.md

```diff
@@ -0,0 +1,124 @@
+3. Add/multisig: No concession is made to `OP_CHECKMULTISIG` or `OP_CHECKSIGADD` semantics with `OP_CHECKSIGFROMSTACK(VERIFY)`. In Tapscript, add semantics can be implemented with 1 additional vByte per key (`OP_TOALTSTACK OP_CHECKSIGFROMSTACK OP_FROMALTSTACK OP_ADD OP_TOALTSTACK`).
+4. Splitting R/S on the stack: Implementing split/separate signatures is left as an exercise for future bitcoin upgrades, such as `OP_CAT`.
+5. BIP118-style Taproot internal key: Rather than introducing an additional key type in this change, we suggest implementing OP_INTERNALKEY or separately introducing that key type for all Tapscript signature checking operations in a separate change.
+6. Unknown key lengths: The semantics of other signature checking opcodes in their respective script types (legacy, segwit-v0, tapscript-c0) are applied.
+
+## Resource Limits
+
+These opcodes are treated identically to other signature checking opcodes and
+count against the various sigops limits and budgets in their respective script
+types.
+
+## Motivation
+
+### LN Symmetry
+
+When combined with `OP_CHECKTEMPLATEVERIFY` (BIP119/CTV),
+`OP_CHECKSIGFROMSTACK` (CSFS) can be used in Lightning Symmetry channels.
+The construction `OP_CHECKTEMPLATEVERIFY <pubkey> OP_CHECKSIGFROMSTACK` is
+logically equivalent to `<bip118_pubkey> OP_CHECKSIG` and a signature over
+`SIGHASH_ALL|SIGHASH_ANYPREVOUTANYSCRIPT`. The `OP_CHECKSIGFROMSTACK`
+construction is 8 vBytes larger.
+
+### Delegation
+
+Using a script like:
+`OP_DUP <pubkey> OP_CHECKSIGFROMSTACK OP_DROP OP_CHECKSIG`
+A script can delegate signing to another key.
+
+## Reference Implementation
+
+A reference implementation is provided in provided here:
+
+https://github.com/brandonblack/bitcoin/commit/5aae0503ceab93101c459748347a111e4a4852c4
+
+## Backward Compatibility
+
+By constraining the behavior of an OP_SUCCESS opcode and an OP_NOP opcode,
+deployment of the BIP can be done in a backwards compatible, soft-fork manner.
+If anyone were to rely on the OP_SUCCESS behavior of
+`OP_SUCCESS204`, `OP_CHECKSIGFROMSTACK` would invalidate
+their spend.
+
+## Deployment
+
+TBD
+
+## Credits
+
+Reference implementation was made with reference to the implementation in
+Elements and started by moonsettler.
+
+## References
+
+[BIP 119](https://github.com/bitcoin/bips/blob/master/bip-0119.mediawiki) CHECKTEMPLATEVERIFY
+
+[BIP 118](https://github.com/bitcoin/bips/blob/master/bip-0119.mediawiki) SIGHASH_ANYPREVOUT
+
+[BIP 340](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki) Schnorr Signatures
+
+[BIP 341](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki) Taproot
+
+[BIP 342](https://github.com/bitcoin/bips/blob/master/bip-0342.mediawiki) Tapscript
+
+[OP_CAT](https://github.com/EthanHeilman/op_cat_draft/blob/main/cat.mediawiki)
+
+## Copyright
+
+This document is placed in the public domain.
```

Comment:
Please add an explicit license; the validity of putting things in the public domain varies across jurisdictions because copyright law is lame. :( cf https://opensource.org/faq/#public-domain